### ISSUE TITLE
WebDav: Exclude / from directory listing

### ISSUE THREAD

**Issue by Flameborn** [drakkan/sftpgo#307]
Currently, SFTPGo sends a / in its directory listing when using the WebDav protocol.

I am not sure what request it sends the list in response to, but my guess is for HEAD only. At any rate, it should not send it neither for *GET* or for *HEAD*. This is usually ignored by clients, so it's not a major issue.

Tested with: a2a99f9

LFTP debug snippet:

```
XML:     <DAV:response>
XML:       <DAV:href>
XML:         `/'
XML:       </DAV:href>
XML:       <DAV:propstat>
XML:         <DAV:prop>
XML:           <DAV:resourcetype>
XML:             <DAV:collection>
XML:             </DAV:collection>
XML:           </DAV:resourcetype>
XML:           <DAV:getlastmodified>
XML:             `Tue, 02 Feb 2021 17:09:14 GMT'
XML:           </DAV:getlastmodified>
XML:           <DAV:supportedlock>
XML:             <DAV:lockentry>
XML:               <DAV:lockscope>
XML:                 <DAV:exclusive>
XML:                 </DAV:exclusive>
XML:               </DAV:lockscope>
XML:               <DAV:locktype>
XML:                 <DAV:write>
XML:                 </DAV:write>
XML:               </DAV:locktype>
XML:             </DAV:lockentry>
XML:           </DAV:supportedlock>
XML:           <DAV:displayname>
XML:           </DAV:displayname>
XML:         </DAV:prop>
XML:         <DAV:status>
XML:           `HTTP/1.1 200 OK'
XML:         </DAV:status>
XML:         `HTTP/1.1 200 OK'
XML:       </DAV:propstat>
XML:       `HTTP/1.1 200 OK'
XML:     </DAV:response>
```

**Comment by drakkan** [comment]
Hi,

are you referring to WebDAV? I just compared the responses from apache `mod_dav` and SFTPGo for a `PROPFIND` request and they are similar, both includes:

```
<?xml version="1.0" encoding="UTF-8"?>
<D:multistatus xmlns:D="DAV:">
   <D:response>
      <D:href>/</D:href>
      .....
```

These seem valid responses according to [RFC 4918](https://tools.ietf.org/html/rfc4918). Any client I tested does not have any issue with these responses. 

Please better specify your use case, including your WebDAV client, and include a network dump or paste the request/response pair that you think is problematic. Thank you

**Comment by Flameborn** [comment]
Hello,

I've edited my first comment to add more detail.

**Comment by drakkan** [comment]
Hi,

I just did a quick test using lftp 4.9.2, the commands listed [here](https://docs.snic.se/wiki/Accessing_Swestore_with_lftp) work fine. Also get/put are ok. ls does not work for me but this seems an lftp issue, it should use PROPFIND and not GET for listing directories contents. 

These settings seems related:

```
hftp:use-allprop
hftp:use-head
hftp:use-mkcol
hftp:use-propfind
```

but they make no difference for me (or I did something wrong): when I type `ls` a GET request is sent and not a PROPFIND, anyway this seems an issue on the lftp side

**Comment by Flameborn** [comment]
Hello,

According to [RFC 4918](http://www.webdav.org/specs/rfc4918.html#rfc.section.9.4), Section 9.4:

> GET, when applied to a collection, may return the contents of an "index.html" resource, a human-readable view of the contents of the collection, or something else altogether.

Based on what SFTPGo returns at the moment, I am assuming that GET for WebDav works similarly, i.e. the response is an XML body.

With a *GET*, and might be via a *HEAD* method as well, listing collections in the root directory is not consistent, since I assume that *propfind* does not return a */* in the listing either, only the children collections.

This is also illustrated if I *cd* into a collection and do a *cls*, or any other listing command via LFTP, the parent directory is consistently not listed, unlike */* in the root. This is why I think that this issue is an inconsistency, regardless of what LFTP uses.

You are right, LFTP is likely trying to access WebDav as a HTTP resource, but the WebDav protocol does not disallow this, as seen above.

**Comment by drakkan** [comment]
Hi,

please also read this from the Section 9.4:

```
Hence, it is possible that the result of a GET on a collection will bear no correlation to the membership of the collection.
``` 

so using GET to list collections content seems wrong. 

Having said that, SFTPGo internally converts GET requests on directory to [PROPFIND with Depth = 1](https://github.com/drakkan/sftpgo/blob/master/webdavd/server.go#L111) and so we return the directory contents. 

I compared the SFTPGo PROPFIND response with apache mod_dav, they are similar, if you list `/` you will get  `<D:href>/</D:href>` and other dirs, if you list `/uploads/` you will get `<D:href>/uploads/</D:href>` and other dirs within uploads.

Here are some network traces:

- `cls` command issued from lftp on `/`, here is the formatted response, as you can see `/` is listed:

```
<?xml version="1.0" encoding="UTF-8"?>
<D:multistatus xmlns:D="DAV:">
   <D:response>
      <D:href>/</D:href>
      <D:propstat>
         <D:prop>
            <D:resourcetype>
               <D:collection />
            </D:resourcetype>
            <D:displayname />
            <D:supportedlock>
               <D:lockentry>
                  <D:lockscope>
                     <D:exclusive />
                  </D:lockscope>
                  <D:locktype>
                     <D:write />
                  </D:locktype>
               </D:lockentry>
            </D:supportedlock>
            <D:getlastmodified>Tue, 09 Feb 2021 08:09:06 GMT</D:getlastmodified>
         </D:prop>
         <D:status>HTTP/1.1 200 OK</D:status>
      </D:propstat>
   </D:response>
   <D:response>
      <D:href>/ros.txt</D:href>
      <D:propstat>
         <D:prop>
            <D:getcontentlength>4034</D:getcontentlength>
            <D:getlastmodified>Tue, 09 Feb 2021 08:09:06 GMT</D:getlastmodified>
            <D:getetag>"166205f06c941c76fc2"</D:getetag>
            <D:getcontenttype>text/plain; charset=utf-8</D:getcontenttype>
            <D:supportedlock>
               <D:lockentry>
                  <D:lockscope>
                     <D:exclusive />
                  </D:lockscope>
                  <D:locktype>
                     <D:write />
                  </D:locktype>
               </D:lockentry>
            </D:supportedlock>
            <D:resourcetype />
            <D:displayname>ros.txt</D:displayname>
         </D:prop>
         <D:status>HTTP/1.1 200 OK</D:status>
      </D:propstat>
   </D:response>
   <D:response>
      <D:href>/adir/</D:href>
      <D:propstat>
         <D:prop>
            <D:getlastmodified>Tue, 09 Feb 2021 08:09:06 GMT</D:getlastmodified>
            <D:resourcetype>
               <D:collection />
            </D:resourcetype>
            <D:displayname>adir</D:displayname>
            <D:supportedlock>
               <D:lockentry>
                  <D:lockscope>
                     <D:exclusive />
                  </D:lockscope>
                  <D:locktype>
                     <D:write />
                  </D:locktype>
               </D:lockentry>
            </D:supportedlock>
         </D:prop>
         <D:status>HTTP/1.1 200 OK</D:status>
      </D:propstat>
   </D:response>
</D:multistatus>
```

and here is the lftp cls output:

```
/  adir/  ros.txt
```

now move inside `adir/`, here is the formatted xml response:

```
<?xml version="1.0" encoding="UTF-8"?>
<D:multistatus xmlns:D="DAV:">
   <D:response>
      <D:href>/adir/</D:href>
      <D:propstat>
         <D:prop>
            <D:getlastmodified>Tue, 09 Feb 2021 08:09:06 GMT</D:getlastmodified>
            <D:supportedlock>
               <D:lockentry>
                  <D:lockscope>
                     <D:exclusive />
                  </D:lockscope>
                  <D:locktype>
                     <D:write />
                  </D:locktype>
               </D:lockentry>
            </D:supportedlock>
            <D:resourcetype>
               <D:collection />
            </D:resourcetype>
            <D:displayname>adir</D:displayname>
         </D:prop>
         <D:status>HTTP/1.1 200 OK</D:status>
      </D:propstat>
   </D:response>
   <D:response>
      <D:href>/adir/Rete%20Bari/</D:href>
      <D:propstat>
         <D:prop>
            <D:supportedlock>
               <D:lockentry>
                  <D:lockscope>
                     <D:exclusive />
                  </D:lockscope>
                  <D:locktype>
                     <D:write />
                  </D:locktype>
               </D:lockentry>
            </D:supportedlock>
            <D:resourcetype>
               <D:collection />
            </D:resourcetype>
            <D:displayname>Rete Bari</D:displayname>
            <D:getlastmodified>Tue, 09 Feb 2021 08:09:06 GMT</D:getlastmodified>
         </D:prop>
         <D:status>HTTP/1.1 200 OK</D:status>
      </D:propstat>
   </D:response>
</D:multistatus>
```

and this is the lftp output

```
Rete Bari/
```

as you can see `/adir/` and `/adir/Rete%20Bari/` are listed within the response but this time lftp only shows `Rete Bari/`. This seems a client side issue

**Comment by Flameborn** [comment]
Hi,

Yes, using *GET* to list collections is definitely not the right way to access a webdav resource.

I see what you mean. I think the issue is that when I reported this, I did not make sure that the output for a non-root directory listing would match a root-directory listing, yet the output would be different.

It seems that this is indeed an LFTP issue. Sorry about the confusion, and thank you very much for giving this a look.
